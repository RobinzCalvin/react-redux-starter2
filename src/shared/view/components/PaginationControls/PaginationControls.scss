.pagination-controls {
  display: flex;
  align-items: center;
  justify-content: center;

  &__arrow {
    width: 0.4rem;
    height: 0.4rem;
    border: solid;
    border-width: 0 0.1rem 0.1rem 0;
    color: #777;
    cursor: pointer;

    &_direction {
      &_left {
        padding-right: 0.5rem;
        transform: rotate(135deg);
      }

      &_right {
        transform: rotate(-45deg);
      }
    }
  }

  &__page {
    margin-right: 0.5rem;
    color: #777;
    cursor: pointer;

    &_active {
      box-shadow: 0 0.1rem;
      color: black;
    }
  }
}
