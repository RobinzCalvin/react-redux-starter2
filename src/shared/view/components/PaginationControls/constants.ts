export const maxRenderedPages = 11;
