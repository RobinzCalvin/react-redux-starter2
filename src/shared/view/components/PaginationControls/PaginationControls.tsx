import React from 'react';
import block from 'bem-cn';
import * as R from 'ramda';
import { bind } from 'decko';

import { maxRenderedPages } from './constants';
import './PaginationControls.scss';

type ArrowDirection = 'left' | 'right';

interface IProps {
  totalPages: number;
  currentPage: number;
  onPageRequest(pageNumber: number): void;
}

const pagesFromMiddlePage = Math.floor(maxRenderedPages / 2);

const b = block('pagination-controls');

class PaginationControls extends React.PureComponent<IProps> {
  public render() {
    const { currentPage, totalPages } = this.props;
    const pagesOnTheLeft = (totalPages - currentPage) < pagesFromMiddlePage
      ? maxRenderedPages - (totalPages - currentPage) - 1
      : pagesFromMiddlePage;
    const firstRenderedPage = Math.max(1, currentPage - pagesOnTheLeft);
    const lastRenderedPage = Math.min(totalPages + 1, (firstRenderedPage + maxRenderedPages));

    return totalPages > 1 && (
      <div className={b()}>
        {this.renderArrow('left', currentPage <= 1)}
        {R.range(firstRenderedPage, lastRenderedPage).map(this.renderPage)}
        {this.renderArrow('right', currentPage >= totalPages)}
      </div>
    );
  }

  private renderArrow(direction: ArrowDirection, hidden: boolean) {
    return <div className={b('arrow', { direction, hidden })} onClick={this.makeArrowClickHandler(direction)}/>;
  }

  @bind
  private renderPage(page: number) {
    const { currentPage } = this.props;
    return (
      <div
        className={b('page', { active: currentPage === page })}
        key={page}
        onClick={this.makePageClickHandler(page)}
      >
        {page}
      </div>
    );
  }

  @bind
  private makePageClickHandler(page: number) {
    return () => this.props.onPageRequest(page);
  }

  @bind
  private makeArrowClickHandler(direction: ArrowDirection) {
    return () => {
      const { onPageRequest, currentPage } = this.props;
      const requestingPage = direction === 'left' ? currentPage - 1 : currentPage + 1;
      onPageRequest(requestingPage);
    };
  }
}

export { IProps as IPaginationControlsProps };
export default PaginationControls;
