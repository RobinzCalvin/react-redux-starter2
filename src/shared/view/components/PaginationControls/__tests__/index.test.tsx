import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import PaginationControls, { IPaginationControlsProps } from '../PaginationControls';

const props: IPaginationControlsProps = {
  totalPages: 5,
  currentPage: 2,
  onPageRequest: jest.fn(),
};

describe('(shared/view) PaginationControls component', () => {
  let component: ShallowWrapper<IPaginationControlsProps>;
  beforeEach(() => component = shallow(<PaginationControls {...props} />));

  it('should call onPageRequest with page number on page number click', () => {
    component.find('.pagination-controls__page').at(2).simulate('click');
    expect(props.onPageRequest).toHaveBeenCalledWith(3);
  });

  it('should call onPageRequest with prev page number on left arrow click', () => {
    const leftArrow = component.find('.pagination-controls__arrow_direction_left');
    leftArrow.simulate('click');
    expect(props.onPageRequest).toHaveBeenCalledWith(props.currentPage - 1);
  });

  it('should call onPageRequest with next page number on right arrow click', () => {
    const rightArrow = component.find('.pagination-controls__arrow_direction_right');
    rightArrow.simulate('click');
    expect(props.onPageRequest).toHaveBeenCalledWith(props.currentPage + 1);
  });

  it('should not render left arrow if first page', () => {
    const leftArrow = component.find('.pagination-controls__arrow_direction_left');
    component.setProps({ currentPage: 1 });
    expect(leftArrow.hasClass('pagination-controls__arrow_hidden')).toBe(false);
  });

  it('should not render right arrow if last page', () => {
    const rightArrow = component.find('.pagination-controls__arrow_direction_right');
    component.setProps({ currentPage: 5 });
    expect(rightArrow.hasClass('pagination-controls__arrow_hidden')).toBe(false);
  });

  it('should not be rendered if totalPages <= 1', () => {
    component.setProps({ totalPages: 1 });
    expect(component.html()).toBeNull();
  });
});
