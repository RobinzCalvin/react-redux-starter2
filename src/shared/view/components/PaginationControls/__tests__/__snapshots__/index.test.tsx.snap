// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationControls component should match snapshot 1`] = `
<PaginationControls
  currentPage={2}
  onPageRequest={[MockFunction]}
  totalPages={5}
>
  <div
    className="pagination-controls"
  >
    <div
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      onClick={[Function]}
    />
    <div
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      key="1"
      onClick={[Function]}
    >
      1
    </div>
    <div
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      key="2"
      onClick={[Function]}
    >
      2
    </div>
    <div
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      key="3"
      onClick={[Function]}
    >
      3
    </div>
    <div
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      key="4"
      onClick={[Function]}
    >
      4
    </div>
    <div
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      key="5"
      onClick={[Function]}
    >
      5
    </div>
    <div
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      onClick={[Function]}
    />
  </div>
</PaginationControls>
`;
